File: dumbwaiter/controller/controller_test.go
package controller

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"

	"github.com/jbruno/dumbwaiter/common"
)

// TestRequestUpFromStop test requesting the car to move up 1 floor when it is stopped
func TestRequestUpFromStop(t *testing.T) {
	// setup
	dwController := setup(t, 2, Stopped, []common.PiPin{common.OpenerUp})

	// test
	dwController.SetRequestedFloor(3)
	waitForStatus(t, 2, 3, Up, dwController, 3*time.Second)
}

// TestRequestStopFromUp test requesting the car to stop floor when it is moving up
func TestRequestArriveAtRequestedFloor(t *testing.T) {
	// setup
	dwController := setup(t, 2, Up, []common.PiPin{common.OpenerStop})

	// test
	dwController.SetRequestedFloor(2)
	waitForStatus(t, 2, 2, Stopped, dwController, 3*time.Second)
}

// TestRequestUpFromMovingDown send an up request when the dumbwaiter is moving down,
// expect to see the dumbwaiter to to stop, then go to moving up
func TestRequestUpFromMovingDown(t *testing.T) {
	// setup
	dwController := setup(t, 2, Down, []common.PiPin{common.OpenerStop, common.OpenerUp})

	// test
	dwController.SetRequestedFloor(3)                       // send the up request
	waitForStatus(t, 2, 3, Up, dwController, 3*time.Second) // verify that the dumbwaiter is now moving up
}

// TestRequestDownFromMovingUp send an down request when the dumbwaiter is moving up,
// expect to see the dumbwaiter to to stop, then go to moving down
func TestRequestDownFromMovingUp(t *testing.T) {
	// setup
	dwController := setup(t, 2, Up, []common.PiPin{common.OpenerStop, common.OpenerDown})

	// test
	dwController.SetRequestedFloor(1)                         // send the down request
	waitForStatus(t, 2, 1, Down, dwController, 3*time.Second) // verify that the dumbwaiter is now moving down
}

// setup creates a controller, and mock pi interface
func setup(t *testing.T, floor int, direction Direction, expectedSendSignals []common.PiPin) *Controller {
	mockRPi := common.NewMockRPi(t, "controllerRPi", expectedSendSignals)
	var dwController *Controller
	dwController = NewController(3).SetRPiDevice(mockRPi).SetLoopFrequency(500 * time.Millisecond)
	dwController.SetLastSeenFloor(2)
	dwController.SetMovingDirection(direction)
	dwController.StartProcessingLoop()
	return dwController
}

func teardown(ticker *time.Ticker) {
	ticker.Stop()
}

// testFloorRequest given the last floor and direction the dumbwaiter is supposed to be moving (lastSeenFloor, currentDirection),
// and the new floor requested (requestedFloor), verify that the dumbwaiter's direction changes to expectedDirection
func testFloorRequest(t *testing.T, lastSeenFloor int, currentDirection Direction, requestedFloor int, expectedDirection Direction, dwc *Controller,
	timingCh chan time.Time) {
	dwc.SetRequestedFloor(requestedFloor)
	timingCh <- time.Now()
	waitForStatus(t, lastSeenFloor, requestedFloor, expectedDirection, dwc, 3*time.Second)
}

func waitForStatus(t *testing.T, lastSeenFloor int, requestedFloor int, expectedDirection Direction, dwc *Controller, timeout time.Duration) {
	waitTill := time.Now().Add(timeout)
	for time.Now().Before(waitTill) {
		s := dwc.GetStatus()
		if expectedDirection == s.MovingDirection && lastSeenFloor == s.LastSeenFloor && requestedFloor == s.RequestedFloor {
			return
		}
		time.Sleep(500 * time.Millisecond)
	}
	s := dwc.GetStatus()
	assert.Equal(t, expectedDirection.String(), s.MovingDirection.String(), "wrong direction")
	assert.Equal(t, requestedFloor, s.RequestedFloor, "wrong requested floor")
	assert.Equal(t, lastSeenFloor, s.LastSeenFloor, "wrong last seen floor")
}

File: dumbwaiter_python/dumbwaiter/tests/controller_test.py
from datetime import datetime, timedelta
from mock import (patch, call)
import time
import unittest

from dumbwaiter.common.rpi import (PiPin, RPiDevice)
from dumbwaiter.controller import (Controller, Direction, default_loop_frequency_sec)

class TestDumbwaiterController(unittest.TestCase):

  @patch('dumbwaiter.common.rpi.RPiDevice.send_signal')
  def test_send_up_from_stopped(self, mock_pi_send):
    dwc = self._setup_test(2, 2, Direction.Stopped)

    mock_pi_send.return_value=None
    dwc.set_requested_floor(3)
    self._wait_for_status(dwc, 2, Direction.Up, 3, 3)
    mock_pi_send.assert_called_once_with(PiPin.OpenerUp)

  @patch('dumbwaiter.common.rpi.RPiDevice.send_signal')
  def test_send_up_from_down(self, mock_pi_send):
    dwc = self._setup_test(2, 1, Direction.Down)

    mock_pi_send.return_value=None
    dwc.set_requested_floor(3)
    self._wait_for_status(dwc, 2, Direction.Up, 3, 3)
    mock_pi_send.assert_has_calls([call(PiPin.OpenerStop), call(PiPin.OpenerUp)])

  @patch('dumbwaiter.common.rpi.RPiDevice.send_signal')
  def test_send_down_from_stopped(self, mock_pi_send):
    dwc = self._setup_test(2, 2, Direction.Stopped)

    mock_pi_send.return_value=None
    dwc.set_requested_floor(1)
    self._wait_for_status(dwc, 2, Direction.Down, 1, 3)
    mock_pi_send.assert_called_once_with(PiPin.OpenerDown)

  @patch('dumbwaiter.common.rpi.RPiDevice.send_signal')
  def test_send_down_from_up(self, mock_pi_send):
    dwc = self._setup_test(2, 3, Direction.Up)

    mock_pi_send.return_value=None
    dwc.set_requested_floor(1)
    self._wait_for_status(dwc, 2, Direction.Down, 1, 3)
    mock_pi_send.assert_has_calls([call(PiPin.OpenerStop), call(PiPin.OpenerDown)])

  # create a controller in the specified state and start it running
  def _setup_test(self, last_seen_floor, requested_floor, direction):
    default_loop_frequency_sec = 0.1  # speed up tests
    dwc = Controller(3)
    dwc.set_last_seen_floor(last_seen_floor)
    dwc.set_requested_floor(requested_floor)
    dwc.set_moving_direction(direction)
    dwc.start_processing_loop()
    return dwc

  # _wait_for_status loops up to the timeout, trying to verify that the controller
  # reached the targeted state
  def _wait_for_status(self, dwc, last_seen_floor, direction, requested_floor, timeout_sec):
    end = datetime.now() + timedelta(seconds=timeout_sec)
    while datetime.now() < end:
      if dwc.get_last_seen_floor() == last_seen_floor and \
        dwc.get_moving_direction() == direction and \
        dwc.get_requested_floor() == requested_floor:
        return
      time.sleep(0.25)

    if dwc.get_last_seen_floor() == last_seen_floor and \
      dwc.get_moving_direction() == direction and \
      dwc.get_requested_floor() == requested_floor:
      return
